### DISCUSSION TITLE
Dirigera mera !!!

### DISCUSSION THREAD

**Discussion by MattWestb** [zigpy/zigpy#1083] [General]
(Eng **Conduct / Direct more !!!**)
I was one day too late then some one have posted photos of it in the shelf in IKEA Wien  Westbahnhof in Reddit forum :-((

I think IKEA have doing one pretty nice hardware but i have not looking under the shields. Only need taking one small torrox screw out from the under side and the gently squeezing the 2 half apart. Then its 2 normal x screws that holding the PCB.
![Dirigera01](https://user-images.githubusercontent.com/49618193/197813292-93521fa1-ea35-4533-ab12-aad79c8eeece.jpg)
![Dirigera02](https://user-images.githubusercontent.com/49618193/197813331-4742ac7c-c938-4a20-9843-33850052bad1.jpg)
2 original Silabs [MGM210LA22JNF2](https://fccid.io/QOQMGM210L/Internal-Photos/Internal-Photos-4211021) with 90° angle to etch other (the center is not the module its the root of the antenna) = by the book !!!  :-))
The 3rd module i dont knowing and but very likely WiFi 2.4 and 5.8 and BT.
And under the shield is likely one router chip with much ram and flash.
Etch Zigbee / Thread module is having one SIlabs Micro (not populated) so full debug and flashing is possible and very likely one comport for the RCP, NCP or the master SOC.

Its looks little like @grobasoz proposed next gen high end coordinator that also have the possibility using more EFR32MG2X chips / modules.
I think our @tube0013 need doing little more cooking for getting one more EFR32MG2X module in his top ZBGW !!!

Next i must installing the new app and testing paring some devices !

**Comment by MattWestb** [comment]
Our @puddly I have sniffing paring devices with Dirigera and its looks very interesting how the is doing it (queering cluster for attributes before doing anything so knowing what the joinee can doing) !!!!
I think Zigbee is calling `find and binding` for getting all thing working without knowing the device before start configure it.

The problem is that it have forming the network on channel 15 then i have 3 NCP with Zigbee and thread network so its very messy in the sniff.

I shall tying deleting the network from the IKEA app and forming one new and see if it making it on one other channel so it being easier looking in the sniff.

You can have the sniff if you like but i sending it then in you e-mail then if i cant  getting changing the network key its bad its going in production have getting support in HA and hd the old hug can going to rest.

**Reply by GSzabados** [reply]
Damn with you @MattWestb, I am just waiting for mine to arrive hopefully tomorrow. You did already what I ordered it for... Good job! 👍

**Reply by MattWestb** [reply]
i was not the first baying it but i but was very curious how IKEA have doing it inside and i is not disappointed how they have doing the hardware as good like they was doing with the generation before (all controller is the hardware made for the EFR32MG1P chips and used optimal but they is from 2019 and have to less hardware resources for thread if making more advanced apps).

Some new 3rd gen light have the antenna inside the shielding power socket and can newer getting the radio working good.

If you is brave and like looking what is under the RF scheduled (with or without destroying the PCB) is photos of it more then welcome !!
In the FCC application is normally both external and internal photos but in this case not and i is not knowing way.

Im curious how this adventure is ending the IKEA have many doors open for doing one great system if they is not doing like OSRAM was doing (killing them self with bad hard and software and loosing the money and the reputation in the end).

PS: I like brave users !!!

**Reply by GSzabados** [reply]
Being the first to buy, we are far behind this guy:

https://www.reddit.com/r/tradfri/comments/xzh3tv/bought_new_dirigera_hub_but_cannot_use_it/

I will see what can be done regarding the RF shield. I am also really curious, but I do not desperately want to damage the device. I want to see how it can be re-purpose d to use with HA without the IKEA environment. 
But it is really not clear for me how the whole thing could be used as a remote gateway from HA with Thread, Bluetooth and Matter.

**Reply by GSzabados** [reply]
By the way, have you seen this comment:

https://community.home-assistant.io/t/new-ikea-dirigera-hub/479895/10

**Reply by MattWestb** [reply]
I know the first one and can using it for weeks :-((

The EFR32MG21 chips is having BT support but im 110% sure not used then many other is cheaper and easier integrating and the box is having one BT chip on the WiFi module with UART connection.
Also having 2 MG21 its possible having one Zigbee network on one channel and one open tread on one other = great.

If having one STM32MP157CAB3 and the rest of the spec its great only if its not read protected.
If its not we can dumping the eMMC and patching it like we have done with LIDL / tuya ZBGW and using all radios over Ethernet in HA for Thread and Zigbee.

**Comment by robertwebbe** [comment]
More pictures here: 
https://www.flickr.com/gp/robertwebbe/f0M0R57uvp

**Reply by MattWestb** [reply]
Nice and much better photos i was doing !!

Have some one looking what for chips is inside the 3 radio module ??

I have not time googling it then im most interested in the Zigbee hardware.

Hoping HA is getting integration for it so can using it in production then i have some outlets and my kitchen sealing light on the old hub with HA automatons that i like braking.

**Reply by MattWestb** [reply]
Ops you have some Eve devices :-))
Is it tread versions ??

**Reply by robertwebbe** [reply]
> Have some one looking what for chips is inside the 3 radio module ??

Murata 2AW wifi module
SS1723068
I couldn't found anything yet, I guess it's quite new.


> Ops you have some Eve devices :-)) Is it tread versions ??

No, old bluetooth versions.

**Reply by MattWestb** [reply]
It can being very new then the Zigbee 3 certificate is for the first / current  hardware version is granted 2021-09-15 https://csa-iot.org/csa_product/dirigera-hub-for-smart-products/ and then the hardware was frozen for production.

I think Murata have doing more WiFi modules but i dont remember if it was Android TV boxes or sticks but im very sure that its not one bad module if they have putting it in the new hub.

Looks like most information that is liked is deleted in google but i have found little in the cash:
```
WLAN: SDIO. 
BT: UART. 
High throughput on 11ac, 
-20°C to +75°C. 
Murata. Type 2AW. (LBEE5ZZ2AW-815). CYW43455. 2.4 / 5GHz, 802.11a/b/g/n. + ac, BT/BLE.
```

**Comment by MattWestb** [comment]
Was deleting the device is the app and disconnecting the app and its saying data is not deleted in the GW. Pressing the reset button under the device for some seconds and its start pulsing for doing init.
Then stop pulsing = its up and running reconnect the app and adding one GU10 WS.
Looking on old ch 15 and its not there and finding it on ch 11 with new network key = good then its not static channel and resetting it is forming on new network = good working !!

Adding one Styrbar and looking how is doing it and its very interesting. Then added the controller to the same room its not controlling the light. Then must clicking on the device "controlled device" for adding the light to control.

Then sniffing it the Styrbar sending unicast command to the coordinator and the coordinator is sending it with group broadcast (looks being not bonded only reseeding the command).

@puddly and @Adminiuga if like see how they is doing the setup with requesting arributes on all cluster before doing reporting and binding i can sending one sniff but i dont like posing it public.

**Comment by GSzabados** [comment]
@wjtje made this nice repo as a collection about the Dirigera Hub:

https://github.com/wjtje/DIRIGERA

**Reply by MattWestb** [reply]
Im interesting how IKEA have doing the LED on the top side of the box.
Its one round circle that is "moving" the loading the system and is full round then all is OK and pulsing then joining device. Also is one point i the middle then its running normal but the point can being green i have seen. I wondering if its one pixel LED with some optic prismas integrated for doing the effects but i cant see how is made on the PCB photos.

I think some sooner or later is trying hacking the GW then the hardware looks nice but im wondering if IKEA have secured the boot so cant changing root password and also changing the U-boot (or EFI bios loader) then the chip have much security futures integrated and can making trouble for getting access to the Linux system.

Also interesting is what OTA URL the new GW is using then IKEA have changing names for over one year for the old GW and i think they is using one new for Dirigera and need being sniffed (it can using the same OTA files but i think the index is being new and also the firmware for the GW and NCP is different).

PS: I was adding one 5 button remote that i have flashed one last good with group binding (with SWD flasher) and it was working but i was seen then sniffing that it was updating the firmware do i was deleting it for not getting it upgraded to new not good firmware.
Perhaps i shall flashing my test OnOff button that is running OpenClose for the moment then the hardware is having all debug connection so its easy fixed is it not working OK.

**Reply by wjtje** [reply]
The led is a ring of white LEDs with one big color led in the middle, I have added a detailed picture in my repo.

The DIRIGERA seems to use the [TF-A BL2](https://wiki.st.com/stm32mpu/wiki/TF-A_BL2_overview) bootloader as indicated by the [log](https://github.com/wjtje/DIRIGERA/blob/main/logs/boot.log) file.

**Reply by MattWestb** [reply]
Thanks @wjtje !!
Your photos is great and also removing the RF shields.
I can see the RAM and the e-MMC but i think its more chips around them. Its looks like 2 flash with normal 8 pins and 2.5 mm pins and also one mini micro ones. Perhaps used for storing custom information or config for the system or temp files for firmware upgrade of the GV and devices connected to it.

I was adding the Dirigera to HA with the `HomeKit Controller` integration and was needing using  the left bar-code but it was not working so was writing the 8 digits with out ` - ` and its was added and the GW and the lighting the middle LED was blinking nice blue for some seconds.

Styrbar is not reconciled but i dont knowing if its only sending events and done have any device entries in HA (luck of experience).

Now we need some crazy hackers for getting inside the bootloader and for root access if its possible.

**Reply by MattWestb** [reply]
Was looking little more on the picture of the LED module and i think its one pixel led with at least 4 white for the ring as one controller chip and one RGB in the middle with controller chip. 
So one LED module with 2 pixel chip one for white ring and one for RGB and need VCC, GND data in and data out and perhaps clock in and out if its using 2 data lines = 4 or 6 line / pads.
And on Robert W last photos its looks having only 4 pads = must being one pixel LED module but not one "normal"  RGB / RGBWW one.

And its looks the community sniffers now can reading the traffic to and from the app then setting up wireshark with sniffing port from one switch https://community.home-assistant.io/t/new-ikea-dirigera-hub/479895/29.
If finding all needed command it using its can being one the starting for one new HA integration but i think we like more low level commands for getting more function out of the hub. Also IKEA is more then likely adding / changing the API then adding more function like the Matter support and more protocols in the (near) future.

But i finding it interesting for the future.


if interested in Matter its being more and more hubs certified for Mater https://csa-iot.org/csa-iot_products/page/9/?p_keywords&p_type%5B0%5D=17&p_type%5B1%5D=14&p_type%5B2%5D=1053&p_program_type%5B0%5D=1049&p_certificate&p_family.

Interesting is Philips HUE and one tuya ZBGW that we have hacked but its using one updated hardware version then our hacked one. I can being that  the Zigbee module was not working (EF32MG2B) was not  able getting RCP working and have changing it to one EFR32MG21 that  i knowing is working OK with  RCP firmware (its only speculations but its looks from outside 110% as my first hacked tuya ZBGW).

**Comment by MattWestb** [comment]
Firmware update is on the  way https://csa-iot.org/csa_product/dirigera-hub-for-smart-products-2/ after more then one year if the initial release firmware.

**Comment by MattWestb** [comment]
## Dirigera OTA files / server and new OTA files?
Diriger is using one unknown  OTA feed and i have trying sniffing  it but i was doing  all wrong :-((
Then adding one new device Dirigera is doing little cloud requests but if its have updating one device with the same file its not downloading it its using the saved cashed file.
Then i have never working with tcpdump its new  land for my.
I connecting to my DD-WRT router and using this command `tcpdump -s 0 host 192.168.2.68 -i br0 -w /mnt/USB-Stick/DG01.pcap` then br0 = the bridge interface so i can getting all in and outgoing traffic and filter on Dirigeras IP and saving the cap on the mounted USB-Stick.

For getting the download information i resetting the Dirigera and adding it in the APP and adding devices that need update. If all is going well its downloading the OTA file  and updating the device.

I dont have the knowledge so if someone is having it pleas share so we can getting updates for our devices.

Im the meantime Zigpy-CLI have getting experimental support for building OTA files from sniffed OTA updates and can using more files is some frames is missing.
I have doing some and posting them in the matrix for IKEA firmware and its looks working great but is mush work (controller is taking over 2 hours only the sniffing),

And the next problem is that Synfonisk 2 is having the debug interface locked so its not possible dumping firmware, bootloader or the user settings in the device, and more then likely is the next gen 2 buttons controllers (OnOff, OpenClose and Shortcut 2) is made in the same way. 

So with this protected device is only one chance sniffing the complete firmware / device and if one frame is missing its not possible reversing  to one older version then IKEA have blocking firmware in there controllers.

If some more users have ides around this pleas post  information so we can getting more out of our devices with or without IKEA native hardware in our systems.

**Reply by GSzabados** [reply]
@MattWestb, I think you are banging on a wide open door....

The old firmware OTA file of the Motion sensor:

`{"fw_binary_url":"http://fw.ota.homesmart.ikea.net/global/GW1.0/01.19.046/bin/10039874-1.0-TRADFRI-motion-sensor-2-2.0.022.ota.ota.signed","fw_file_version_LSB":9763,"fw_file_version_MSB":8194,"fw_filesize":186814,"fw_image_type":4552,"fw_manufacturer_id":4476,"fw_type":2}`

The new firmware OTA file of the Motion sensor and Remote Control:

```
{"fw_binary_url":"http://fw.ota.homesmart.ikea.net/global/GW1.0/01.19.046/bin/159699-5.1-TRADFRI-remote-control-24.4.5.ota.ota.signed","fw_file_version_LSB":5,"fw_file_version_MSB":9220,"fw_filesize":207084,"fw_image_type":4545,"fw_manufacturer_id":4476,"fw_type":2}
{"fw_binary_url":"http://fw.ota.homesmart.ikea.net/global/GW1.0/01.19.046/bin/159700-TRADFRI-motion-sensor-24.4.5.ota.ota.signed","fw_file_version_LSB":5,"fw_file_version_MSB":9220,"fw_filesize":214316,"fw_image_type":4548,"fw_manufacturer_id":4476,"fw_type":2}
```

Source for all of these:

http://fw.ota.homesmart.ikea.net/feed/version_info.json

This is the original OTA feed from IKEA.

**Comment by MattWestb** [comment]
Yes they is in TF OTA but im missing at least 5 that Dirigera have updated:
<html>
<body>


OTA File Name | Image Type Hex/Dec | Product String in FW | HW gen / Zigbee ver | TRÅDFRI | DIRIGERA
-- | -- | -- | -- | -- | --


</body>
</html>
<html>
<body>


symfonisk_sound_remote_zingo | 0x110E 4366 | SYMFONISK sound remote 2 (E2123) | 3 ZB3 | 🔲 | 1.0.32 ☑️
-- | -- | -- | -- | -- | --

<html>
<body>


tradfri_connected_blind | 0x1187 4487 | TRÅDFRI KADRILJ roller blind (E1752-140) | 2 ZB3 | 2.3.087 ✅ | 24.4.10 ☑️
-- | -- | -- | -- | -- | --

<html>
<body>


tradfri_onoff_shortcut_contr | 0x11C5 4549 | TRÅDFRI on/off switch (E1743) | 1 ZLL > ZB3 | 2.3.079 ✅ | 24.4.5 ☑️
-- | -- | -- | -- | -- | --
tradfri_shortcut_button | 0x11C6 4550 | TRADFRI SHORTCUT Button (E1812) | 2 ZB3 | 2.3.080 ✅ | 24.4.5 ☑️
<html>
<body>


tradfri_dimmer | 0x11CA 4554 | SYMFONISK Sound Controller (E1744) | 2 ZB3 | 2.3.080 ✅ | 24.4.5 ☑️
-- | -- | -- | -- | -- | --


And  very likely new sensors is not being released in TF feed then its not supporting them and need Dirigera.

The old motion sensor have also getting one 24.X update but i have not making any OTA file for it then i need re flashing it with one old  FW and sniffing the upgrading. and building the  signed OTA file with zigpy-CLI.

**Reply by GSzabados** [reply]
OK, sorry, I though, you were looking for that two. By the way, why the motion sensor has a new fw_image_type? Z2M cannot pick it up....

**Reply by MattWestb** [reply]
The old or the new version of it ?
Its 2 different devices with different OTA images (but the false name inside the firmware).
Z2M is using the TF feed and you is only getting updates for it.
You can downloading the OTA files from Dirigera and installing them with Z2M  but its little  more work.

**Reply by GSzabados** [reply]
It is the new one, or newer one, with the 3 buttons on the back, not the old one with the adjust dial.

**Reply by GSzabados** [reply]
I actually see what is the issue. they left the new version's OTA in the file twice. 

10039874-1.0-TRADFRI-motion-sensor-2-2.0.022.ota.ota.signed is current firmware for the motion sensor.

159700-TRADFRI-motion-sensor-1.2.214.ota.ota.signed and this one should be the firmware for the old motion sensor. (This second one is from the the TEST OTA Feed.)

**Reply by MattWestb** [reply]
The IKEA test feed is DEAD as of October 2021 after they was having bad files that was doing controller not being able to upgrade.
The old motion was the last getting Zigbee 3 firmware 25th Jan 2023 with `159700-TRADFRI-motion-sensor-24.4.5.ota.ota.signedand` the new was getting the last OTA 2.0.022 for some year and i have no release notes of it but my local file is from 2021 and the date in the file is 201903008.
Likely is coming on update for the new one in TF feed in the next release or IKEA is holding it back for not braking tush link paring in TF.

**Reply by GSzabados** [reply]
Oh, ok, then it explains it. I though that they updated the current, newer version, of the Motion sensor, and not the old one.

**Reply by MattWestb** [reply]
Nop they was doing **all** devices Zigbee 3 compatible and working with better with Dirigera.

**Comment by rrakso** [comment]
Hi all! I'm wondering is it possible to flash the Dirigera with a custom firmware? Something like a lightweight version of a Home Assistant?

**Reply by MattWestb** [reply]
Very possible but it have locked boot-loader so not easy getting your firmware inside it.
The firmware source code is posted online so can getting information of how they is using the hardware.
But no one have hacking it and its little expensive for trashing but  if you is  good programming STM  chips think its very interesting hardware to getting one real Linux running on that can being used with HA.

**Comment by dylangerdaly** [comment]
Has anyone managed to dump firmware? I'm going to remove the THGBMNG5D1LBAIL from the board but need the datasheet, can't find it anywhere online - I'm guessing some of the unpopulated debug pads on the board go directly to the eMMC PINs, but need the datasheet to know what specific PINs.

The GPL releases that IKEA release don't include their Cloud Connect/OTA userland binaries, this is what we'd mostly be interested in for getting newer firmware from their OTA Server.

**Comment by dylangerdaly** [comment]
I managed to dump the eMMC running on the Dirigera, check out my post here

https://withblack.ink/posts/fun-with-ikeas-dirigera/fun-with-ikeas-dirigera/

Comes with the updated OTA URL.

**Reply by MattWestb** [reply]
## Great work done !!
But desolating the eMMC and reading it with brute force is on the the boarder to sheeting but was working well.
Shall updating the new OTA feed URL in the IKEA OTA discussion .

**Reply by GSzabados** [reply]
> is on the the boarder to sheeting but was working well.

Nah, it isn't cheating. It was a fair effort. And a job well done.

But now IKEA might have learnt that they should encrypt their partitions.

**Reply by MattWestb** [reply]
Sorry for then but now its too late for IKEA devs doing that !!
If i remember OK is the main SOC having hard and software support for securing its storage inclusive the secure boot but if not implanting it then . . . 
They was doing the same with there 3 gen devices was shipped un locked chip / secure storage and later start locking all device from the factory.
I was dumping the bootloader from one (un locked) Styrbar and flashing it on one MG21 module and extracting the OTA for Symfonisk II and flashing it on the APP part of the flash and it was working OK and also OTA updated with the unlocked chip inclusive secure storage of private keys was unlocked.

By the way have your Dirigera change the color of the white LED to ember ?
Its looks like the LED is burning out and changing color then being old.

**Reply by dylangerdaly** [reply]
I didn't know this before I posted but the OTA link was already public lol

If you guys are wanting to run your own OS on it, it's possible, you'd need to buy a fresh STM32MP157CAB3 SoC and replace it, from there you can boot firmware over the USB.

IKEA are using Fastboot over the USB-C port to provision devices on the factory floor

You can actually see exactly how IKEA are doing this [here](https://github.com/ikea-oss-distributions/DIRIGERA-hub/blob/main/u-boot-stm32mp-v2020.10-stm32mp-r2-r0/0014-Fuse-root-key-hash.patch#L60-L94), they're blowing a SHA-256 hash into BSEC_OTP_RKH at provisioning time.

Because I'd like to sharpen my glitching skillz a little when I have some free time I'm going to see if I can bypass this check with either Vcc glitching or EMFI, once there I can load and run a verison of optee that'll print the private key material for the TLS Certs

I doubt IKEA will start encrypting firmware, too much effort for basically no gain.

After some more research, if you replace the main SoC with an unfused one, nothing in IKEA's stock firmware blows anything, it'll come up normally, you can then use the IKEA's OSS Dump to build your own firmware, what would break is IKEA App TLS Authentication.

Technically someone could come up with a total OSS version of firmware with no IKEA related cloud things, and their own controller daemons, but that's a lot of work.

Another cool little find is in U-Boot, if you can manage to write to the `TAMP_BKP20R` register, you can force the device to boot into an eMMC Storage Device, meaning you can re-program/dump the eMMC over the USB port, this code exists in the production version of the device - possibly via rebooting with `reboot ums_mmc0` from the device or `busybox devmem 0x5C00A150 32 0x00000010`

Also the LED is totally controlled by the M4 ARM Co-Processor, no idea why IKEA did this, possibly becuase it wasn't being used for anything and RTOS ensures LEDs run smoothly

These are the commands Linux can pass to the M4 Co-Processor

```
  ┌───────────────────────────────────────────┬───────────────────────────────────────────────────────────────────────────────┐                                           
  │                  Command                  │                                    Purpose                                    │
  ├───────────────────────────────────────────┼───────────────────────────────────────────────────────────────────────────────┤
  │ M4Cmd_LED_SET_OUTPUT                      │ Raw on/off of a specific LED channel                                          │
  ├───────────────────────────────────────────┼───────────────────────────────────────────────────────────────────────────────┤
  │ M4Cmd_LED_SET_RGB                         │ Set RGB triplet on a specific LED                                             │                                           
  ├───────────────────────────────────────────┼───────────────────────────────────────────────────────────────────────────────┤                                           
  │ M4Cmd_LED_EFFECTS_ENABLE                  │ Start a named animation effect                                                │                                           
  ├───────────────────────────────────────────┼───────────────────────────────────────────────────────────────────────────────┤                                           
  │ M4Cmd_LED_EFFECTS_DISABLE                 │ Stop a named effect                                                           │
  ├───────────────────────────────────────────┼───────────────────────────────────────────────────────────────────────────────┤                                           
  │ M4Cmd_LED_EFFECTS_CHANGE_PARAMS           │ Live-tune the current effect (brightness, speed, ramp)                        │
  ├───────────────────────────────────────────┼───────────────────────────────────────────────────────────────────────────────┤                                           
  │ M4Cmd_LED_EFFECTS_WAIT_FOR_ITER_END       │ Synchronous wait: block until the running effect completes one cycle          │
  ├───────────────────────────────────────────┼───────────────────────────────────────────────────────────────────────────────┤                                           
  │ M4Cmd_LED_EFFECTS_BOOTUP_PATTERN_FINISHED │ Signal from A7 that early-boot UI can now hand off to whatever effect is next │
  └───────────────────────────────────────────┴───────────────────────────────────────────────────────────────────────────────┘        
```

**Reply by MattWestb** [reply]
I was trying sniffing with wireshark the URL for OTA but it was using TLS encrypted to the cloud so was dumping sniffs and building the OTA files and it was working  OK if getting all packages.
Then one user was diving in the source code IKEA released and getting the first URL and later we have the same as you was finding and its used in ZHA for long time now for doing OTA updates of IKEA devices.
IKEAs Zigbee devices is not encrypted but is being signed so the device is doing one check after have receiving the file and if its OK its flashing it and if not its throwing it

I agree with IKEA is not start locking down the hardware then its very expensive (not only € but more reputation) and not worth it for them and dont  try to fix somthing that is not broken most is saying but **THAT is boring !!!**.
Hope is going well with your EMFI glitching then i have not looking if the STM32 is easy getting out of its cage and getting one nice root privilege on the not modified hardware.

**Reply by dylangerdaly** [reply]
Ah you mean IKEA's ZigBee end devices? Like the switches/bulbs etc? These are signed? What modifications would you like to make?

I'm trying to get better at soldering/desoldering etc, I'm going to buy 2 test Dirigera's, if I manage to successfully swap out the SoC I'll send you one ;)

Keep in mind EMFI/Glitching is only useful as a one-shot thing, you can boot with debug SWD enabled and dump keys, in a single session

After poking around IKEA's U-Boot bootloader and OP-TEE, I don't see any RMA style fuses, these usually allow vendors to blow a prod unit back to debug/EVT state - the only true way to run your own firmware is to replace the SoC with a clean unfused one.

**Reply by MattWestb** [reply]
Some Danish friends is / was doing custom wall switches with reusing the MG21 module from Styrbar dimmer switch and it was not debug locked so easy dumping the flash and user data burned in it. Then they was baying sheepest IKEA light bulb for taking the module out and reusing it. Then IKEA star debug locking the modules but can being erased and loosing all the flash but userdata is not so only writing old bootloader and one "app" and all was working OK. They have dumping all devices that was getting there hands on so can easy converting devices if like. IKEA was using standard Silabs SDK but with there own custom application and some custom patches with 2 different working mode so not easy moding if not  having very deep knowledge in ARM desambling.

Burning Dirigeras is not so bad if being in EU and mot of the word but in USA is around 100 USD with the extra for china import there have implemented.

**Reply by dylangerdaly** [reply]
I just took apart a spare motion sensor I had laying around, it looks like there are clearly labelled PINs for SWD, I'm pretty sure you can load whatever firmware you want over the PINs usually behind the batteries.

I'm surprised at how much logic can be put into a MG21.

I managed to pull 2 of the MGM210L from this dead Dirigera so it's not all wasted

**Reply by MattWestb** [reply]
Both the first and second gens MG1P and 3thd gen MG21 devices is very easy only need one SWD probe and for MG2X one that can handling the Sirlabs extra functionality. I have not looking on Vallhorn if the early and the later produced one is debug locked or not if being locked you need do one chip erase / debug unlock for getting contact with the internal flash (if not glitching the chip that shall being possible then its only software locked then the hardware locked chips is more expensive). If its not its very easy dumping all flash / memory. The main flash have first the bootloader and then the "app" = one copy of the OTA file plus NVM at the upper end. Then the second flash part named "user data" have the chip and costumer settings like the device name "VALLHORN Wireless Motion Sensor" radio calibration and so on and its not erased by debug unlocking !!

The MGM210L is one very good Zigbee / Thread module and IKEA was getting it very cheep from Silabs but its little locked down if compared normal MG21 chips but then its pre  FCC and CE certified.

You can find the pining IKEA have using on my Billy / Markus Git and also flash layouts if you like using the module for other things or putting some working code on them (I have working Zigbee NCP for it but cant getting Thread working OK).

If you was having Thread working on the Dirigera and can dumping the flash of the MGM210L im interested of the Thread RCP firmware they is using then i have not getting it working with Silabs SDK on the MGM210L.

**Reply by dylangerdaly** [reply]
The firmware for both Thread and Zigbee (MGM210L) are located in `/lib/firmware/silabs`
```
> $ ls                                                                                                                                                                   
ot-rcp-zingo.gbl  zigbee_ncp.bin  zigbee_ncp.s37  zigbee_se.bin
```
The Thread firmware is `ot-rcp-zingo.gbl`

Looking at strings it appears it's a fork of OpenThread

`SL-OPENTHREAD/2.4.5.0_GitHub-797150858; EFR32; Jenkins`

Someone should 3D Print a pogo pin setup for MG21 as it's all just debugging over serial/SWD, I started poking around with Zigbee with the nRF52840 MDK USB Dongle, once you have your own NWK its neat being able to sniff all of the messages.

**Reply by MattWestb** [reply]
Sounds great then i must taking one lock in the hub firmware !
You can also taking one first seond gen MG1P module or one MG21 and loading it with standard Zigbee NCP firmware and using it as one sniffer only one USB.TTL after flashing one UART bootloader.
Mr **Jenkins** we have seen in all 3rd gen firmware so hi is the Silabs COOCKER and the tag _zingo_ looks being the code name for all MG21 devices in there firmware.
